### PR TITLE
[clang][Interp] Handle CXXScalarValueInitExprs

### PR DESCRIPTION
**Pull request by tbaederr**

### CODE REVIEW

File: clang/test/AST/Interp/literals.cpp [hunk 1]
```diff
@@ -31,6 +31,13 @@ static_assert(b, "");
 constexpr int one = true;
 static_assert(one == 1, "");
 
+constexpr bool b2 = bool();
+static_assert(!b2, "");
+constexpr int ScalarInitInt = int();
+static_assert(ScalarInitInt == 0, "");
+constexpr float ScalarInitFloat = float();
+static_assert(ScalarInitInt == 0.0f, "");
```
Comment (tbaederr) [inline clang/test/AST/Interp/literals.cpp:39]:
```suggestion
static_assert(ScalarInitFloat == 0.0f, "");
```

**Review by erichkeane — APPROVED** [review]
Implementation looks simple/right enough.  LGTM.

File: clang/test/AST/Interp/literals.cpp [hunk 2]
```diff
@@ -31,6 +31,13 @@ static_assert(b, "");
 constexpr int one = true;
 static_assert(one == 1, "");
 
+constexpr bool b2 = bool();
```
Comment (shafik) [inline clang/test/AST/Interp/literals.cpp:34]: There are a lot of scalars: https://eel.is/c++draft/basic.types.general#9

Can we have a test for each case?

**Review by shafik — COMMENTED** [review]
Looks good but I want more tests.

File: clang/test/AST/Interp/literals.cpp [hunk 3]
```diff
@@ -31,6 +31,28 @@ static_assert(b, "");
...
+
+  template<typename T>
+  constexpr T getScalar() { return T(); }
+
+  static_assert(getScalar<const int>() == 0, "");
+  static_assert(getScalar<const double>() == 0.0, "");
+
+  static_assert(getScalar<void*>() == nullptr, "");
+  static_assert(getScalar<void(*)(void)>() == nullptr, "");
+  /// FIXME: Member pointers.
```
Comment (shafik) [inline clang/test/AST/Interp/literals.cpp:53]: and enumeration types

**Review by shafik — COMMENTED** [review]
Thank you for the additions, almost there :-)

**Comment by tbaederr** [comment]
I've removed the second commit, so this needs https://github.com/llvm/llvm-project/pull/67229 to be merged first.

File: clang/test/AST/Interp/literals.cpp [hunk 4]
```diff
@@ -31,6 +31,33 @@ static_assert(b, "");
...
+  static_assert(getScalar<const double>() == 0.0, "");
+
+  static_assert(getScalar<void*>() == nullptr, "");
+  static_assert(getScalar<void(*)(void)>() == nullptr, "");
+
+  enum E {
+    First = 0,
+  };
+  static_assert(getScalar<E>() == First, "");
+  /// FIXME: Member pointers.
```
Comment (shafik) [inline clang/test/AST/Interp/literals.cpp:58]: So we are leaving out member pointers b/c we will take care of it on a subsequent PR?

**Review by shafik — APPROVED** [review]

File: clang/test/AST/Interp/literals.cpp (hunk 4 above)
Comment (tbaederr) [reply clang/test/AST/Interp/literals.cpp:58]: Yes, because member pointers just don't work at all right now.

**Comment by tbaederr** [comment]
Ping

**Review by shafik — APPROVED** [review]
LGTM, just fill out the description with some more details so folks reading git log will have more context,

**Pull request merged**